File: src/main/java/org/taktik/icure/logic/impl/DocumentTemplateLogicImpl.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.logic.impl;

import java.util.Collection;
import java.util.List;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.taktik.icure.dao.DocumentTemplateDAO;
import org.taktik.icure.entities.DocumentTemplate;
import org.taktik.icure.entities.embed.DocumentType;
import org.taktik.icure.logic.DocumentTemplateLogic;
import org.taktik.icure.logic.ICureSessionLogic;

@Service
public class DocumentTemplateLogicImpl extends GenericLogicImpl<DocumentTemplate, DocumentTemplateDAO> implements DocumentTemplateLogic {
	private static Logger logger = LoggerFactory.getLogger(DocumentTemplateLogicImpl.class);

	private DocumentTemplateDAO documentTemplateDAO;
	private ICureSessionLogic sessionLogic;

	@Autowired
	public void setDocumentTemplateDAO(DocumentTemplateDAO documentTemplateDAO) {
		this.documentTemplateDAO = documentTemplateDAO;
	}

	@Override
	public boolean createEntities(Collection<DocumentTemplate> entities, Collection<DocumentTemplate> createdEntities) throws Exception {
		entities.stream().forEach((e) -> {
			if (e.getOwner() == null) {
				e.setOwner(sessionLogic.getCurrentUserId());
			}
		});
		return super.createEntities(entities, createdEntities);
	}


	@Override
	public DocumentTemplate createDocumentTemplate(DocumentTemplate entity) {
		if (entity.getOwner() == null) {
			entity.setOwner(sessionLogic.getCurrentUserId());
		}

		return documentTemplateDAO.createDocumentTemplate(entity);
	}

	@Override
	public DocumentTemplate getDocumentTemplateById(String documentTemplateId) {
		return documentTemplateDAO.get(documentTemplateId);
	}

	@Override
	public List<DocumentTemplate> getDocumentTemplatesBySpecialty(String specialityCode) {
		return documentTemplateDAO.findBySpecialtyGuid(specialityCode, null);
	}

	@Override
	public List<DocumentTemplate> getDocumentTemplatesByDocumentType(String documentTypeCode) {
		return documentTemplateDAO.findByTypeUserGuid(documentTypeCode, null, null);
	}

	@Override
	public List<DocumentTemplate> getDocumentTemplatesByDocumentTypeAndUser(String documentTypeCode,String userId) {
		return documentTemplateDAO.findByTypeUserGuid(documentTypeCode, userId, null);
	}

	@Override
	public List<DocumentTemplate> getDocumentTemplatesByUser(String userId) {
		return documentTemplateDAO.findByUserGuid(userId, null);
	}

	@Override
	public DocumentTemplate modifyDocumentTemplate(DocumentTemplate documentTemplate) {
		if (documentTemplate.getOwner() == null) {
			documentTemplate.setOwner(sessionLogic.getCurrentUserId());
		}

		return documentTemplateDAO.save(documentTemplate);
	}

	@Override
	protected DocumentTemplateDAO getGenericDAO() {
		return documentTemplateDAO;
	}

	@Autowired
	public void setSessionLogic(ICureSessionLogic sessionLogic) {
		this.sessionLogic = sessionLogic;
	}
}

File: src/main/kotlin/org/taktik/icure/asynclogic/impl/DocumentTemplateLogicImpl.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.asynclogic.impl

import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.emitAll
import kotlinx.coroutines.flow.flow
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.taktik.icure.asyncdao.DocumentTemplateDAO
import org.taktik.icure.asynclogic.AsyncSessionLogic
import org.taktik.icure.asynclogic.DocumentTemplateLogic
import org.taktik.icure.entities.DocumentTemplate


@ExperimentalCoroutinesApi
@Service
class DocumentTemplateLogicImpl(private val documentTemplateDAO: DocumentTemplateDAO,
                                private val sessionLogic: AsyncSessionLogic) : GenericLogicImpl<DocumentTemplate, DocumentTemplateDAO>(sessionLogic), DocumentTemplateLogic {

    override fun createEntities(entities: Collection<DocumentTemplate>): Flow<DocumentTemplate> = flow {
        emitAll(super.createEntities(
                entities.map {dt ->
                    fix(dt) { e ->
                        e.owner?.let { e } ?: e.copy(owner = sessionLogic.getCurrentUserId())
                    }
                }))
    }

    override suspend fun createDocumentTemplate(documentTemplate: DocumentTemplate) = fix(documentTemplate) { documentTemplate ->
        documentTemplateDAO.createDocumentTemplate(documentTemplate.owner?.let { documentTemplate } ?: documentTemplate.copy(owner = sessionLogic.getCurrentUserId()))
    }

    override suspend fun getDocumentTemplate(documentTemplateId: String): DocumentTemplate? {
        return documentTemplateDAO.get(documentTemplateId)
    }

    override fun getDocumentTemplatesBySpecialty(specialityCode: String): Flow<DocumentTemplate> = flow {
        emitAll(documentTemplateDAO.findBySpecialtyGuid(specialityCode, null))
    }

    override fun getDocumentTemplatesByDocumentType(documentTypeCode: String): Flow<DocumentTemplate> = flow {
        emitAll(documentTemplateDAO.findByTypeUserGuid(documentTypeCode, null, null))
    }

    override fun getDocumentTemplatesByDocumentTypeAndUser(documentTypeCode: String, userId: String): Flow<DocumentTemplate> = flow {
        emitAll(documentTemplateDAO.findByTypeUserGuid(documentTypeCode, userId, null))
    }

    override fun getDocumentTemplatesByUser(userId: String): Flow<DocumentTemplate> = flow {
        emitAll(documentTemplateDAO.findByUserGuid(userId, null))
    }

    override suspend fun modifyDocumentTemplate(documentTemplate: DocumentTemplate) = fix(documentTemplate) { documentTemplate ->
        documentTemplateDAO.save(documentTemplate.owner?.let { documentTemplate } ?: documentTemplate.copy(owner = sessionLogic.getCurrentUserId()))
    }

    override fun getGenericDAO(): DocumentTemplateDAO {
        return documentTemplateDAO
    }

    companion object {
        private val logger = LoggerFactory.getLogger(DocumentTemplateLogicImpl::class.java)
    }
}
